browser/browsing_data/brave_clear_browsing_data.cc

```diff
@@ -0,0 +1,182 @@
+
+  // Note: this will also delete Brave Shields site-specific settings.
+  // Corresponds to "Content settings" checkbox in the Clear Browsing Data
+  // dialog.
+  if (prefs->GetBoolean(browsing_data::prefs::kDeleteSiteSettingsOnExit))
+    *remove_mask |=
+        ChromeBrowsingDataRemoverDelegate::DATA_TYPE_CONTENT_SETTINGS;
+
+  return (*remove_mask != 0);
+}
+
+// This method will, for each loaded profile that is not off-the-record, gather
+// the user specified types of data that need to be cleared. It will then get
+// the BrowsingDataRemover for that profile and call its RemoveAndReply method.
+// BrowsingDataRemover will create a number of tasks to clear the data. Because
+// these tasks, on their own, will neither prevent shutdown nor stop the profile
+// from being destroyed, we have to block shutdown execution from proceeding any
+// further. Otherwise the tasks will be cancelled and the profiles destroyed.
+// Since we can't actually block the UI thread, instead we implement the Wait
+// method below, which just runs a RunLoop. When a BrowsingDataRemover finishes
+// its tasks it will reply back to us by calling the OnBrowsingDataRemoverDone
+// method below. When that happens we decrement the counter of profiles that
+// need to be cleared. Once the counter reaches 0 we exit the RunLoop and let
+// shutdown proceed.
+void BrowsingDataRemovalWatcher::ClearBrowsingDataForLoadedProfiles(
+    BraveClearBrowsingData::OnExitTestingCallback* testing_callback) {
+  ProfileManager* profile_manager = g_browser_process->profile_manager();
+  DCHECK(profile_manager);
+
+  std::vector<Profile*> profiles = profile_manager->GetLoadedProfiles();
+  for (Profile* profile : profiles) {
+    if (profile->IsOffTheRecord())
+      continue;
+    int remove_mask;
+    int origin_mask;
+    if (!GetClearBrowsingDataOnExitSettings(profile, &remove_mask,
+                                            &origin_mask))
+      continue;
+    ++num_profiles_to_clear_;
+    content::BrowsingDataRemover* remover =
+        content::BrowserContext::GetBrowsingDataRemover(profile);
+    observer_.Add(remover);
+    if (testing_callback)
+      testing_callback->BeforeClearOnExitRemoveData(remover, remove_mask,
+                                                    origin_mask);
+    remover->RemoveAndReply(base::Time(), base::Time::Max(), remove_mask,
+                            origin_mask, this);
+  }
+
+  Wait();
+}
+
+void BrowsingDataRemovalWatcher::Wait() {
+  if (num_profiles_to_clear_ > 0)
+    run_loop_.Run();
+}
+
+void BrowsingDataRemovalWatcher::OnBrowsingDataRemoverDone() {
+  --num_profiles_to_clear_;
+  if (num_profiles_to_clear_ > 0)
+    return;
+
+  run_loop_.Quit();
+}
+
+}  // namespace
+
+namespace content {
+
+BraveClearBrowsingData::OnExitTestingCallback*
+    BraveClearBrowsingData::on_exit_testing_callback_ = nullptr;
+
+// static
+void BraveClearBrowsingData::ClearOnExit() {
+  BrowsingDataRemovalWatcher watcher;
```

Comment:
Nit: I'd suggest to put `TRACE_EVENT0` here and there, in case of future performance investigations. This is also a good place to put `UMA_HISTOGRAM_*` measurements, but we do not collect them yet.

Reply:
Put a single one for now for the entire operation. More can be added as needed.